pkg/controllers/machinemigration/machine_migration_controller.go

```diff
@@ -0,0 +1,492 @@
+	} else if removed {
+		// Wait for it to take effect.
+		return ctrl.Result{RequeueAfter: time.Second}, nil
+	}
+
+	logger.Info("Switching authoritativeAPI for machine")
+
+	// Set the actual AuthoritativeAPI to the desired one, reset the synchronized generation and condition.
+	if err := r.setNewAuthoritativeAPIAndResetSynchronized(ctx, mapiMachine, mapiMachine.Spec.AuthoritativeAPI); err != nil {
+		return ctrl.Result{}, fmt.Errorf("unable to apply authoritativeAPI to status with patch: %w", err)
+	}
+
+	logger.Info("Successfully switched authoritativeAPI for machine")
+
+	// Make sure the new authoritative resource has been requested to unpause.
+	if err := r.ensureUnpauseRequestedOnNewAuthoritativeResource(ctx, mapiMachine); err != nil {
+		return ctrl.Result{}, fmt.Errorf("unable to ensure the new AuthoritativeAPI has been un-paused: %w", err)
+	}
+
+	logger.Info("Successfully unpaused new authoritative machine")
+	logger.Info("Migration completed for machine")
+
+	return ctrl.Result{}, nil
+}
+
+// isOldAuthoritativeResourcePaused checks whether the old authoritative resource is paused.
+func (r *MachineMigrationReconciler) isOldAuthoritativeResourcePaused(ctx context.Context, m *machinev1beta1.Machine) (bool, error) {
+	if m.Spec.AuthoritativeAPI == machinev1beta1.MachineAuthorityClusterAPI {
+		cond, err := util.GetConditionStatus(m, "Paused")
+		if err != nil {
+			return false, fmt.Errorf("unable to get paused condition for %s/%s: %w", m.Namespace, m.Name, err)
+		}
+
+		return cond == corev1.ConditionTrue, nil
+	}
+
+	// For MachineAuthorityMachineAPI, check the corresponding CAPI resource.
+	capiMachine := &capiv1beta1.Machine{}
+	if err := r.Get(ctx, client.ObjectKey{Namespace: consts.DefaultManagedNamespace, Name: m.Name}, capiMachine); err != nil {
+		return false, fmt.Errorf("failed to get Cluster API machine: %w", err)
+	}
+
+	machinePausedCondition := conditionsv1beta2.Get(capiMachine, capiv1beta1.PausedV1Beta2Condition)
+	if machinePausedCondition == nil {
+		return false, nil
+	}
+
+	infraMachineRef := capiMachine.Spec.InfrastructureRef
+
+	infraMachine, err := util.GetReferencedObject(ctx, r.Client, r.Scheme, infraMachineRef)
+	if err != nil {
+		return false, fmt.Errorf("failed to get Cluster API infra machine: %w", err)
+	}
+
+	infraMachinePausedConditionStatus, err := util.GetConditionStatus(infraMachine, capiv1beta1.PausedV1Beta2Condition)
+	if err != nil {
+		return false, fmt.Errorf("unable to get paused condition for %s/%s: %w", infraMachine.GetNamespace(), infraMachine.GetName(), err)
+	}
+
+	return (machinePausedCondition.Status == metav1.ConditionTrue) && (infraMachinePausedConditionStatus == corev1.ConditionTrue), nil
+}
+
+func (r *MachineMigrationReconciler) ensureUnpauseRequestedOnNewAuthoritativeResource(ctx context.Context, mapiMachine *machinev1beta1.Machine) error {
+	// Request that the new authoritative resource reconciliation is un-paused.
+	//nolint:wsl,exhaustive
```

Comment:
Why do these not matter?

Reply:
`exhaustive`

The only other entry for the switch on `mapiMachine.Spec.AuthoritativeAPI` is Migrating, which is disallowed to be set at the spec level, only at the status level (I covered it with a comment in `default`)

https://github.com/openshift/api/blob/8fcc4e71758aefadc7bf7da355ca8619bcc1d34d/machine/v1beta1/types_machine.go#L269


`wsl`
is complaining because some `case` are empty, but I want to explicitly keep them so we can add the comment there.